lib/matplotlib/widgets.py

```diff
@@ -1237,35 +1250,108 @@ def _get_data(self, event):
+
+        self._prev_event = event
+        return event
+
     def press(self, event):
-        """Button press handler"""
+        """Button press handler and validator"""
         if not self.ignore(event):
-            self.eventpress = copy.copy(event)
-            self.eventpress.xdata, self.eventpress.ydata = (
-                self._get_data(event))
+            event = self._clean_event(event)
+            self.eventpress = event
+            self._prev_event = event
+            key = event.key or ''
+            key = key.replace('control', 'ctrl')
+            if key == self.state_modifier_keys['move']:
+                self.state.add('move')
+            self._press(event)
             return True
         return False
 
+    def _press(self, event):
+        """Button press handler"""
+        pass
+
     def release(self, event):
-        """Button release event"""
-        if not self.ignore(event) and self.eventpress is not None:
-            self.eventrelease = copy.copy(event)
-            self.eventrelease.xdata, self.eventrelease.ydata = (
-                self._get_data(event))
+        """Button release event handler and validator"""
+        if not self.ignore(event) and self.eventpress:
+            event = self._clean_event(event)
+            self.eventrelease = event
+            self._release(event)
+            self.eventpress = None
+            self.eventrelease = None
+            self.state.discard('move')
             return True
-        else:
-            return False
+        return False
+
+    def _release(self, event):
+        """Button release event handler"""
+        pass
 
     def onmove(self, event):
-        """Cursor move event"""
+        """Cursor move event handler and validator"""
+        if not self.ignore(event) and self.eventpress:
+            event = self._clean_event(event)
+            self._onmove(event)
+            return True
+        return False
+
+    def _onmove(self, event):
+        """Cursor move event handler"""
         pass
 
     def on_scroll(self, event):
-        """Mouse scroll event"""
+        """Mouse scroll event handler and validator"""
+        if not self.ignore(event):
+            self._on_scroll(event)
+
+    def _on_scroll(self, event):
+        """Mouse scroll event handler"""
         pass
 
     def on_key_press(self, event):
-        """Key press event"""
+        """Key press event handler and validator"""
+        if self.active:
+            key = event.key or ''
+            key = key.replace('control', 'ctrl')
+            if key == self.state_modifier_keys['clear']:
+                for artist in self.artists:
+                    artist.set_visible(False)
+                self.update()
+                return
+            for (state, modifier) in self.state_modifier_keys.items():
+                if modifier in key:
+                    self.state.add(state)
+            self._on_key_press(event)
```

Comment:
I added some docs here.  Essentially the `_` methods are for Widget-specific things, and the top level is meant for all Selection widgets.  Clear is not so much as state as an action - clear the current selection.